api/metrics/v1/rules.go

```diff
@@ -0,0 +1,239 @@
+package v1
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/go-chi/chi"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/rulefmt"
+	"gopkg.in/yaml.v3"
+
+	"github.com/observatorium/api/authentication"
+)
+
+// ErrRuleNotFound is returned when a particular rule wasn't found by its name.
+var ErrRuleNotFound = errors.New("rule not found")
+
+type RuleGroups struct {
+	Groups []RuleGroup `yaml:"groups"`
+}
+
+type RuleGroup struct {
+	Name     string         `yaml:"name"`
+	Interval model.Duration `yaml:"interval"`
+	Rules    []rulefmt.Rule `yaml:"rules"`
+}
+
+// RulesRepository describes all of the operations that a conformant repository should implement.
+type RulesRepository interface {
+	RulesLister
+	RulesGetter
+	RulesWriter
+}
+
+type RulesLister interface {
+	ListRuleGroups(ctx context.Context, tenant string) (RuleGroups, error)
+}
+
+func listRulesHandler(logger log.Logger, lister RulesLister, label string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		tenant, ok := authentication.GetTenant(r.Context())
+		if !ok {
+			http.Error(w, "failed to get tenant", http.StatusInternalServerError)
+			return
+		}
+
+		id, ok := authentication.GetTenantID(r.Context())
+		if !ok {
+			const msg = "error finding tenant ID"
+			level.Warn(logger).Log("msg", msg)
+			http.Error(w, msg, http.StatusInternalServerError)
+			return
+		}
+
+		rules, err := lister.ListRuleGroups(r.Context(), tenant)
+		if err != nil {
+			msg := "failed to list rules"
+			level.Debug(logger).Log("msg", msg, "err", err)
+			http.Error(w, msg, http.StatusInternalServerError)
+			return
+		}
+
+		for i := range rules.Groups {
+			for j := range rules.Groups[i].Rules {
+				if rules.Groups[i].Rules[j].Labels == nil {
+					rules.Groups[i].Rules[j].Labels = make(map[string]string)
```

Comment:
Why do we need to do this? To prevent a panic when accessing presumably?

Reply:
yes exactly, we're not guaranteed to have a map allocated yet